### PR TITLE
Softmax + cross Entropy fusion for numerical Stabilization

### PR DESCRIPTION
**Pull request by pruthviIntel**
This PR implements Softmax + CrossEntropy as single layer for numerical stabilization.

image

The second equation in the above image is the vanilla implementation but suffers from numerical instability, the Dataflow graph for the same is as below.

image

We fuse the numerical in stabilized graph and replace with the DataFlow graph which implements the last equation, DFG for the same after graph replacement is as below. 

image

### **Softmax + Crossentropy Gradient**

Reduces the equation to 
image
for numerical stabilization when soft_lables is enabled from bridge

Before Optimization

image

After Optimization

image

### CODE REVIEW

File: src/ngraph/runtime/cpu/cpu_external_function.cpp [hunk 1]
```diff
@@ -1203,6 +1203,8 @@ void runtime::cpu::CPU_ExternalFunction::register_common_passes(
     };
 
     REGISTER_KNOBBED_PASS(LikeReplacement, true, ngraph::pass)
+    REGISTER_KNOBBED_PASS_WITH_ARGS(
+        CoreFusion, true, ngraph::pass, ngraph::pass::FusionType::ALL_FUSIONS)
```
Comment (jbobba) [inline src/ngraph/runtime/cpu/cpu_external_function.cpp]: Instead of moving CoreFusion up, add a second FusedOpDecomposition after CoreFusion.

File: src/ngraph/pass/core_fusion.cpp [hunk 2]
```diff
@@ -55,6 +59,83 @@ static shared_ptr<Node> construct_constant_node(int n)
...
+    auto multiply = std::make_shared<ngraph::op::Multiply>(broadcast_2, subtract);
+    auto bprop_result = std::make_shared<ngraph::op::Result>(multiply);
+
+    auto callback = [constant_1_label, constant_2_label, constant_3_label, param_1, param_2](
+        pattern::Matcher& m) {
+        NGRAPH_DEBUG << "In a callback for construct_sigmoid_cross_entropy_bprop against "
+                     << m.get_match_root()->get_name();
+
+        auto pattern_map = m.get_pattern_map();
+
```
Comment (jbobba) [inline src/ngraph/pass/core_fusion.cpp]: Add a TODO.

File: src/ngraph/op/fused/softmax_crossentropy.hpp [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+                                const AxisSet& reduction_axes);
+
+            virtual NodeVector decompose_op() const override;
+
+            virtual std::shared_ptr<Node>
+                copy_with_new_args(const NodeVector& new_args) const override;
+
+            const AxisSet& get_summation_axis() const { return m_summation_axis; }
+        private:
+            AxisSet m_summation_axis;
```
Comment (jbobba) [inline src/ngraph/op/fused/softmax_crossentropy.hpp]: Either use 'reduction_axes' or 'summation_axes' consistently throughout.

File: test/cpu_fusion.cpp [hunk 4]
```diff
@@ -4012,3 +4012,47 @@ NGRAPH_TEST(cpu_fusion, backwards_batchmatmultranspose_tensor2_tensor2)
 #endif
 
 #endif
+
```
Comment (jbobba) [inline test/cpu_fusion.cpp]: These tests need to be guarded by NGRAPH_JSON_DISABLE. They will also need to be moved to core_fusion.

**Comment by pruthviIntel** [comment]
@jbobba can u take a look again, i have addressed your feed back

File: src/ngraph/op/fused/softmax_crossentropy.cpp [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+#include "ngraph/op/subtract.hpp"
+#include "ngraph/op/sum.hpp"
+#include "ngraph/op/util/broadcasting.hpp"
+
+using namespace std;
+using namespace ngraph;
+
+constexpr NodeTypeInfo op::SoftmaxCrossEntropy::type_info;
+
+op::SoftmaxCrossEntropy::SoftmaxCrossEntropy(const Output<Node>& arg1,
```
Comment (baojun-nervana) [inline src/ngraph/op/fused/softmax_crossentropy.cpp]: I see softmaxCrossEntroy fprop. How can I call softmaxCrossEntropy bprop from bridge.

Comment (pruthviIntel) [reply src/ngraph/op/fused/softmax_crossentropy.cpp]: I am adding a separate PR for bprop.

File: src/ngraph/serializer.cpp [hunk 6]
```diff
@@ -3336,6 +3343,12 @@ json JSONSerializer::serialize_node(const Node& n)
         }
         break;
     }
+    case OP_TYPEID::SoftmaxCrossEntropy:
+    {
+        auto tmp = dynamic_cast<const op::SoftmaxCrossEntropy*>(&n);
```
Comment (jbobba) [inline src/ngraph/serializer.cpp]: static_cast

File: test/core_fusion.cpp [hunk 7]
```diff
@@ -702,3 +702,93 @@ TEST(batch_fusion, pass_property)
     ASSERT_EQ(true, pass->get_property(pass::PassProperty::REQUIRE_STATIC_SHAPE));
     ASSERT_EQ(false, pass->get_property(pass::PassProperty::CHANGE_DYNAMIC_STATE));
 }
+
+#ifndef NGRAPH_JSON_DISABLE
```
Comment (mbrookhart) [inline test/core_fusion.cpp]: Is this defined or not defined by default?

Comment (pruthviIntel) [reply test/core_fusion.cpp]:
```
option(NGRAPH_FAST_MATH_ENABLE "Enable fast math" ON)
option(NGRAPH_JSON_ENABLE "Enable JSON based serialization and tracing features" TRUE)
```
its JOSN is enabled by default

File: src/ngraph/op/fused/softmax_crossentropy.cpp [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+                                             int64_t reduction_axes,
+                                             bool soft_label,
+                                             int64_t ignore_index)
+    : FusedOp({arg1, arg2})
+    , m_soft_label(soft_label)
+    , m_ignore_index(ignore_index)
+{
+    if (reduction_axes < 0)
+    {
+        m_reduction_axes = arg1.get_shape().size() + reduction_axes;
```
Comment (baojun-nervana) [inline src/ngraph/op/fused/softmax_crossentropy.cpp]: This won't work for dynamic shape. You may check this PR: https://github.com/NervanaSystems/ngraph/pull/3708/files

File: src/ngraph/op/fused/softmax_crossentropy.cpp [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+
+NodeVector op::SoftmaxCrossEntropy::decompose_op() const
+{
+    auto input_to_normalize = input_value(0);
+    auto labels = input_value(1);
+
+    // always reduces the sum on the last axis
+    auto reduction_axis = input_to_normalize.get_shape().size() - 1;
+    auto max_xj =
+        std::make_shared<ngraph::op::Max>(input_to_normalize, AxisSet{size_t(reduction_axis)});
```
Comment (baojun-nervana) [inline src/ngraph/op/fused/softmax_crossentropy.cpp]: use static_cast?

File: src/ngraph/op/fused/softmax_crossentropy.cpp [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+    auto broadcast_max_xj =
+        std::make_shared<ngraph::op::Broadcast>(max_xj, input_to_normalize.get_shape(), AxisSet{1});
+    auto subtract = std::make_shared<ngraph::op::Subtract>(input_to_normalize, broadcast_max_xj);
+    auto exp = std::make_shared<ngraph::op::Exp>(subtract);
+
+    auto sum_over_j = std::make_shared<ngraph::op::Sum>(exp, AxisSet{size_t(reduction_axis)});
```
Comment (baojun-nervana) [inline src/ngraph/op/fused/softmax_crossentropy.cpp]: static_cast?

File: src/ngraph/op/fused/softmax_crossentropy.cpp [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+
+    auto subtract_max_xj_from_input =
+        std::make_shared<ngraph::op::Subtract>(input_to_normalize, broadcast_max_xj);
+    auto broadcast_log = std::make_shared<ngraph::op::Broadcast>(
+        log_sum_over_j, subtract_max_xj_from_input->get_shape(), AxisSet{1});
+    auto subtract_max_xj_from_input_from_log_sum_over_j =
+        std::make_shared<ngraph::op::Subtract>(subtract_max_xj_from_input, broadcast_log);
+    auto multiply = std::make_shared<ngraph::op::Multiply>(
+        labels, subtract_max_xj_from_input_from_log_sum_over_j);
+    auto sum_over_k = std::make_shared<ngraph::op::Sum>(multiply, AxisSet{size_t(reduction_axis)});
```
Comment (baojun-nervana) [inline src/ngraph/op/fused/softmax_crossentropy.cpp]: static_cast?

File: src/ngraph/op/fused/softmax_crossentropy.cpp [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+    auto labels = input_value(2);
+    size_t one_hot_axis = delta.get_shape().size() - 1;
+    // always reduces the sum on the last axis
+    auto reduction_axis = delta.get_shape().size() - 1;
+
+    if (m_soft_label)
+    {
+        auto delta_mul_labels = std::make_shared<ngraph::op::Multiply>(delta, labels);
+        auto summation_delta_mul_labels =
+            std::make_shared<ngraph::op::Sum>(delta_mul_labels, AxisSet{size_t(reduction_axis)});
```
Comment (baojun-nervana) [inline src/ngraph/op/fused/softmax_crossentropy.cpp]: static_cast

File: src/ngraph/op/fused/softmax_crossentropy.cpp [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+        auto convert_one_hot = std::make_shared<ngraph::op::Convert>(one_hot, element::f64);
+
+        // (cross_entr * delta * mask)
+        auto delta_mul_labels = std::make_shared<ngraph::op::Multiply>(delta, convert_one_hot);
+        auto multiply_mask =
+            std::make_shared<ngraph::op::Multiply>(delta_mul_labels, broadcast_mask);
+
+        // sum (cross_entr * delta * mask)
+        auto summation_delta_mul_labels =
+            std::make_shared<ngraph::op::Sum>(multiply_mask, AxisSet{size_t(reduction_axis)});
```
Comment (baojun-nervana) [inline src/ngraph/op/fused/softmax_crossentropy.cpp]: static_cast?

File: src/ngraph/op/fused/softmax_crossentropy.hpp [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+            /// \param labels Node that produces ground truth labels for input
+            /// \param soft_label flag indicating whether to interpretate the given labels as soft
+            /// labels
+            /// \param ignore_index Specifies a target value that is ignored and does not contribute
+            /// to the input gradient Only valid if soft_label is set to False
+            SoftmaxCrossEntropyBackprop(const Output<Node>& delta,
+                                        const Output<Node>& softmax,
+                                        const Output<Node>& labels,
+                                        bool soft_label = false,
+                                        int64_t ignore_index = -100);
```
Comment (baojun-nervana) [inline src/ngraph/op/fused/softmax_crossentropy.hpp]: we may not need this for bprop.
